File: c_files/raylib/pong.c
#include <raylib.h>
#include <raymath.h>

// BALL

typedef struct Ball
{
  Vector2 initial_pos;
  Vector2 pos;
  float radius;
  float speed;
  Vector2 direction;
  Color color;
  bool active;

} Ball;

void ball_init(Ball *ball)
{
  ball->initial_pos =
      (Vector2){GetScreenWidth() * 0.5f, GetScreenHeight() * 0.5f};
  ball->pos = ball->initial_pos;
  ball->radius = 10.0f;
  ball->speed = 400.0f;
  ball->direction = (Vector2){1, 1};
  ball->color = RED;
  ball->active = false;
}

void ball_reset(Ball *ball)
{
  ball->pos = ball->initial_pos;
  ball->direction.x *= -1;
  ball->active = false;
}

void ball_draw(Ball *ball)
{
  DrawCircleV(ball->pos, ball->radius, ball->color);
}

void ball_update(Ball *ball)
{
  // activate/deactivate ball
  if (IsKeyPressed(KEY_SPACE))
  {
    ball->active = !ball->active;
  }

  // move ball
  if (ball->active)
  {
    ball->pos.x += ball->direction.x * ball->speed * GetFrameTime();
    ball->pos.y += ball->direction.y * ball->speed * GetFrameTime();
  }

  // ball bounds
  if (ball->pos.y < ball->radius ||
      ball->pos.y > GetScreenHeight() - ball->radius)
  {
    ball->direction.y *= -1;
  }

  // normalize direction vector
  if (ball->direction.x != 0 && ball->direction.y != 0)
  {
    ball->direction = Vector2Normalize(ball->direction);
  }
}

// PADDLE

typedef struct Paddle
{
  Rectangle initial_rect;
  Rectangle rect;
  float speed;
  Color color;

} Paddle;

void paddle_init(Paddle *paddle, float x)
{
  float height = 100.0f;
  paddle->initial_rect =
      (Rectangle){x, GetScreenHeight() * 0.5f - height / 2, 10.0f, height};
  paddle->rect = paddle->initial_rect;
  paddle->speed = 300.0f;
  paddle->color = BLACK;
}

void paddle_reset(Paddle *paddle) { paddle->rect = paddle->initial_rect; }

void paddle_draw(Paddle *paddle)
{
  DrawRectangleRec(paddle->rect, paddle->color);
}

void paddle_update_player(Paddle *player)
{
  // move paddle
  if (IsKeyDown(KEY_UP))
  {
    player->rect.y -= player->speed * GetFrameTime();
  }
  if (IsKeyDown(KEY_DOWN))
  {
    player->rect.y += player->speed * GetFrameTime();
  }

  // paddle bounds
  player->rect.y =
      Clamp(player->rect.y, 0, GetScreenHeight() - player->rect.height);
}

void paddle_update_ai(Paddle *ai, Ball *ball)
{
  // move paddle
  if (ball->pos.y < ai->rect.y + ai->rect.height / 2)
  {
    ai->rect.y -= ai->speed * GetFrameTime();
  }
  if (ball->pos.y > ai->rect.y + ai->rect.height / 2)
  {
    ai->rect.y += ai->speed * GetFrameTime();
  }

  // paddle bounds
  ai->rect.y = Clamp(ai->rect.y, 0, GetScreenHeight() - ai->rect.height);
}

// GAME MANAGER

typedef struct Game
{
  int score_left;
  int score_right;

  Ball ball;
  Paddle player;
  Paddle ai;

} Game;

void game_init(Game *game)
{
  game->score_left = 0;
  game->score_right = 0;

  ball_init(&game->ball);
  paddle_init(&game->player, 10.0f);
  paddle_init(&game->ai, GetScreenWidth() - game->player.rect.width - 10.0f);
}

void game_reset(Game *game)
{
  ball_reset(&game->ball);
  paddle_reset(&game->player);
  paddle_reset(&game->ai);
}

void game_draw(Game *game)
{
  // draw scores
  DrawText(TextFormat("%d", game->score_left), 200, 30, 40, BLACK);
  DrawText(TextFormat("%d", game->score_right), GetScreenWidth() - 200, 30, 40,
           BLACK);

  ball_draw(&game->ball);
  paddle_draw(&game->player);
  paddle_draw(&game->ai);
}

void game_update(Game *game)
{
  ball_update(&game->ball);
  paddle_update_player(&game->player);
  paddle_update_ai(&game->ai, &game->ball);

  // ball collision paddle
  if (CheckCollisionCircleRec(game->ball.pos, game->ball.radius,
                              game->player.rect))
  {
    game->ball.direction.x *= -1;
  }
  if (CheckCollisionCircleRec(game->ball.pos, game->ball.radius,
                              game->ai.rect))
  {
    game->ball.direction.x *= -1;
  }

  // update scores
  if (game->ball.pos.x < game->ball.radius)
  {
    game->score_right += 1;
    game_reset(game);
  }
  if (game->ball.pos.x > GetScreenWidth() - game->ball.radius)
  {
    game->score_left += 1;
    game_reset(game);
  }
}

// MAIN

int main(void)
{
  const int SCREEN_WIDTH = 800;
  const int SCREEN_HEIGHT = 600;
  const char SCREEN_TITLE[] = "Pong";
  const Color SCREEN_BACKGROUND = SKYBLUE;

  InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE);

  Game game;
  game_init(&game);

  while (!WindowShouldClose())
  {

    BeginDrawing();
    ClearBackground(SCREEN_BACKGROUND);

    game_update(&game);
    game_draw(&game);

    EndDrawing();
  }

  CloseWindow();

  return 0;
}

File: python_files/raylib/pong.py
import pyray as p

SCREEN_WIDTH: int = 800
SCREEN_HEIGHT: int = 600
SCREEN_TITLE: str = "Pong"
SCREEN_BACKGROUND: p.Color = p.SKYBLUE


class Ball:
    def __init__(self) -> None:
        self.initial_x: float = p.get_screen_width() / 2
        self.initial_y: float = p.get_screen_height() / 2
        self.x: float = self.initial_x
        self.y: float = self.initial_y
        self.radius: float = 10.0
        self.speed: float = 400.0
        self.direction: p.Vector2 = p.Vector2(
            -1 if p.get_random_value(0, 1) == 0 else 1,
            -1 if p.get_random_value(0, 1) == 0 else 1,
        )
        self.color: p.Color = p.RED
        self.active: bool = False

    def reset(self) -> None:
        self.x = self.initial_x
        self.y = self.initial_y
        self.direction.x *= -1
        self.active = False

    def draw(self) -> None:
        p.draw_circle_v(p.Vector2(self.x, self.y), self.radius, self.color)

    def update(self) -> None:
        # activate ball
        if p.is_key_pressed(p.KeyboardKey.KEY_SPACE):
            self.active = not self.active

        if abs(self.pos.x) > 0.01 and abs(self.pos.y) > 0.01:
            self.direction = p.vector2_normalize(self.direction)

        # move ball
        if self.active:
            self.x += self.direction.x * self.speed * p.get_frame_time()
            self.y += self.direction.y * self.speed * p.get_frame_time()

        # ball bounds
        if self.y < self.radius or self.y > p.get_screen_height() - self.radius:
            self.direction.y *= -1


class Paddle:
    def __init__(self, x: float) -> None:
        self.width: float = 10.0
        self.height: float = 100.0
        self.initial_x: float = x
        self.initial_y: float = p.get_screen_height() / 2 - self.height / 2
        self.x: float = self.initial_x
        self.y: float = self.initial_y
        self.speed: float = 300.0
        self.color: p.Color = p.BLACK

    def reset(self) -> None:
        self.x = self.initial_x
        self.y = self.initial_y

    def draw(self) -> None:
        p.draw_rectangle_rec(
            p.Rectangle(self.x, self.y, self.width, self.height), self.color
        )

    def update_player(self) -> None:
        # move paddle
        if p.is_key_down(p.KeyboardKey.KEY_UP):
            self.y -= self.speed * p.get_frame_time()
        if p.is_key_down(p.KeyboardKey.KEY_DOWN):
            self.y += self.speed * p.get_frame_time()

        # paddle bounds
        self.y = max(0.0, min(self.y, p.get_screen_height() - self.height))

    def update_ai(self, ball: Ball) -> None:
        # move paddle
        if ball.y < self.y + self.height / 2:
            self.y -= self.speed * p.get_frame_time()
        if ball.y > self.y + self.height / 2:
            self.y += self.speed * p.get_frame_time()

        # paddle bounds
        self.y = max(0.0, min(self.y, p.get_screen_height() - self.height))


class Game:
    def __init__(self) -> None:
        self.score_left: int = 0
        self.score_right: int = 0

        self.ball: Ball = Ball()
        self.player: Paddle = Paddle(10.0)
        self.ai: Paddle = Paddle(p.get_screen_width() - self.player.width - 10.0)

    def reset(self) -> None:
        self.ball.reset()
        self.player.reset()
        self.ai.reset()

    def draw(self) -> None:
        # draw scores
        p.draw_text(str(self.score_left), 200, 30, 40, p.BLACK)
        p.draw_text(str(self.score_right), p.get_screen_width() - 200, 30, 40, p.BLACK)

        self.ball.draw()
        self.player.draw()
        self.ai.draw()

    def update(self) -> None:
        self.ball.update()
        self.player.update_player()
        self.ai.update_ai(self.ball)

        # ball collision paddle
        if p.check_collision_circle_rec(
            p.Vector2(self.ball.x, self.ball.y),
            self.ball.radius,
            p.Rectangle(
                self.player.x, self.player.y, self.player.width, self.player.height
            ),
        ):
            self.ball.direction.x *= -1

        if p.check_collision_circle_rec(
            p.Vector2(self.ball.x, self.ball.y),
            self.ball.radius,
            p.Rectangle(self.ai.x, self.ai.y, self.ai.width, self.ai.height),
        ):
            self.ball.direction.x *= -1

        # update scores
        if self.ball.x < self.ball.radius:
            self.score_right += 1
            self.reset()
        if self.ball.x > p.get_screen_width() - self.ball.radius:
            self.score_left += 1
            self.reset()


def main() -> None:
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE)

    game: Game = Game()

    while not p.window_should_close():
        p.begin_drawing()
        p.clear_background(SCREEN_BACKGROUND)

        game.draw()
        game.update()

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()
